### PR TITLE
Identity based encryption

### PR DESCRIPTION
**Pull request by 3pmile**
**Description**

This PR implements...
- [x] DualRegevIBE (with gen, extract, enc, dec)
- [x] new, new_from_n and default for DualRegevIBE
- [x] check_correctness, check_security
- [x] IBE

for/ of IBE.

**Testing**

- [x] I added basic working examples (possibly in doc-comment)
- [x] I triggered all possible errors in my test in every possible way (except for check_correctness and check_security)
- [x] I included tests for all reasonable edge cases
- [x] I provided an intuition regarding how certain inputs have to be set

**Checklist:**

- [x] I have performed a self-review of my own code
  - [x] The code provides good readability and maintainability s.t. it fulfills best practices like talking code, modularity, ...
    - [x] The chosen implementation is not more complex than it has to be
  - [x] My code should work as intended and no side effects occur (e.g. memory leaks)
  - [x] The doc comments fit our style guide
  - [x] I have credited related sources if needed

### CODE REVIEW

File: src/construction/identity_based_encryption/dual_regev_ibe.rs [hunk 1]
```diff
@@ -345,7 +349,7 @@ impl IBE for DualRegevIBE {
         let secret_key = self.psf.samp_p(master_pk, master_sk, &u);
 
         // insert secret key in HashMap
-        self.storage.insert(identity.clone(), secret_key.clone()); //todo insert for different pk and sk
+        self.storage.insert(identity.clone(), secret_key.clone());
```
Comment (jnsiemer) [inline src/construction/identity_based_encryption/dual_regev_ibe.rs:352]: Just insert a tuple `(master_pk, master_sk, identity)` as key into the hash map (or if this is getting too big, use a hashed version of the tuple) to get rid of the WARNING.

**Review by jnsiemer — CHANGES_REQUESTED** [review]

File: src/construction/identity_based_encryption/dual_regev_ibe.rs (hunk 1 above)
Comment (3pmile) [reply src/construction/identity_based_encryption/dual_regev_ibe.rs]: applyed

File: src/construction/identity_based_encryption/dual_regev_ibe.rs [hunk 2]
```diff
@@ -562,4 +564,19 @@ mod test_dual_regev_ibe {
...
+    fn extract_storage_same_identity_different_mk_pk() {
+        let id = String::from(format!("Hello World!"));
+        let mut cryptosystem = DualRegevIBE::default();
+        let (pk_1, sk_1) = cryptosystem.setup();
+        let (pk_2, sk_2) = cryptosystem.setup();
+
+        let id_sk_1 = cryptosystem.extract(&pk_1, &sk_1, &id);
+        let id_sk_2 = cryptosystem.extract(&pk_2, &sk_2, &id);
+
+        assert!(id_sk_1 != id_sk_2)
```
Comment (jnsiemer) [inline src/construction/identity_based_encryption/dual_regev_ibe.rs:580]: ```assert_ne!(id_sk_1, id_sk_2)```
clippy should report this :)

**Review by jnsiemer — APPROVED** [review]
